baseapp/baseapp.go

```diff
@@ -718,35 +720,85 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 	// Result if any single message fails or does not have a registered Handler.
 	result, err = app.runMsgs(runMsgCtx, msgs, mode)
 
-	if err == nil {
+	// Case 1: the msg errors and the post handler is not set.
+	if err != nil && app.postHandler == nil {
+		return gInfo, nil, events, priority, err
+	}
 
-		// Run optional postHandlers.
-		//
-		// Note: If the postHandler fails, we also revert the runMsgs state.
-		if app.postHandler != nil {
-			// Follow-up Ref: https://github.com/cosmos/cosmos-sdk/pull/13941
-			newCtx, err := app.postHandler(runMsgCtx, tx, mode == runTxModeSimulate, err == nil)
-			if err != nil {
-				return gInfo, nil, anteEvents, priority, err
-			}
+	// Case 2: tx errors and the post handler is set. Run PostHandler and revert state from runMsgs
+	if err != nil && app.postHandler != nil {
```

Comment:
Is there consensus on running postHandler _after_ a failed tx? I believe @amaurym had objections to this.

Reply:
This change is the motivation for this PR (including the change from #13940 which @amaurym approved).

The idea is that now the post-handler decorators can specify if they only run when the tx succeeds or fails based on the new `success` field.